### ISSUE TITLE
Lack of deferred deeplink exposed call

### ISSUE THREAD

**Issue by juanimoli** [adjust/android_sdk#425]
Hi,

We're making a library for our application in which many deferred deeplink proivders can be added for fetching upon app's first start. For that we are managing the threading by ourselves. We're trying to add Adjust, but we cannot find an easy way to fetch that deeplink. The main idea is to fetch all providers asynchronously (each separately in parallel) in other thread (rather than UI/main) and wait all them to finish.

For this, we're looking to encapsulate all fetching logic in a simple way to the developer, so that with a simple line you can get a deferred deeplink. A quick example could be [AppLinkData Facebook's deferred implementation](https://developers.facebook.com/docs/reference/android/current/class/AppLinkData/?locale=es_ES) in which with a simple call to **fetchDeferredAppLinkData**, we can get the deeplink. I'm seeing that Adjust's SDK has the concept of callback or listener, but lacks of a simple method to make the call whenever we want (correct me if I'm wrong). 

Is there any way to initialize Adjust (like we're doing right now) and fetching deferred deeplink later?

If not, is something you want/you have in backlog?

Thanks!

**Comment by uerceg** [comment]
Hi @juanimoli.

Interesting project, cool idea. 👍 

What you elaborated and noticed is pretty much correct - we will fire deferred deep link callback if implemented and other than that, there's no dedicated getter for its content. There are few reasons why this getter doesn't exist. First most obvious one is that SDK is not doing local persistence of it to provide it at some later point. And second one is actually the reason why we don't persist it locally - information about deferred deep link will be provided to SDK after install of an app happens, but install how Adjust is defining it. So in theory, if install happened with deferred deep link attached, user reinstalls an app and upon next SDK start after reinstall, we will actually not get the deferred deep link information at this point since reinstall is not being treated as an install on our end. It's just how things currently work.

We're always opened for ideas and suggestions of this kind, so feel free to fire request for this feature to either support@adjust.com or your dedicated AM if you have one and it will reach our product team to be evaluated.

In case you have any further questions on this topic, feel free to ping.

Cheers

**Comment by juanimoli** [comment]
Hey @uerceg !

Thanks for your response and for your explanation. I pretty much got the idea, but I still have a question.

What if Adjust makes all these checks before making the call? That would make Adjust not firing API calls when it's not first install (and we still have a deferred deeplink response, but a null one) and would give apps the possibility to fire it whenever they want. 

I didn't really check if this is possible and how detection of first install vs reinstall is being made, but I'm pretty sure for some reason it needs to be done upon app's start. My proposal would be to still make that check upon app's start, but provide an internally way to check this later (as in a field in AdjustManager or something like that) so that Adjust can check this value when any app tries to fetch a deferred. This could also separate responsabilities between Adjust initialization and redirection.

Of course it's just an idea, but I'm glad to be heared! Thanks

**Comment by uerceg** [comment]
Hey @juanimoli 

Thanks for your comments! Adjust SDK can't make these checks prior to talking to our backend because SDK can't make this distinction based on any criteria we have at our disposal locally in the app. Backend is at the end of the day the only source of this information.

We are however planning to something about it in next major SDK release which might help this situation. Plans we have is to add `deeplink` parameter into `AdjustAttribution` object (where it also naturally belongs in our backend logic) and then once attribution callback is triggered after an app install, if this field is `null`, it will indicate that the install happened without any deep link attached to tracker URL. If it has value, deferred deep link information will be delivered in here.

Also in addition to that, we're probably gonna add additional field to `AdjustAttribution` object called `source` (or similar) which will indicate inside of the attribution callback was it triggered after what our backend saw as an install or is it just some of the subsequent app reinstalls.
